hdk/jack/syntax.py

```diff
@@ -0,0 +1,614 @@
+
+    Attributes:
+        type_: The type of the variable.
+        names: The names of the variables.
+    """
+
+    type_: str
+    names: list[str]
+
+    @property
+    def is_identifier(self) -> bool:
+        """True if type is identifier."""
+        return self.type_ not in {"int", "char", "boolean"}
+
+    def to_xml(self, doc: Document) -> Element:
+        element = doc.createElement("varDec")
+        tag = "identifier" if self.is_identifier else "keyword"
+        _add_children(element, ("keyword", "var"), (tag, self.type_))
+        for i, name in enumerate(self.names):
+            _add_children(element, ("identifier", name))
+            if i < len(self.names) - 1:
+                _add_child(element, ",")
+        _add_child(element, ";")
+        return element
+
+    def __post_init__(self):
+        if not _is_symbol_valid(self.type_):
+            raise ValueError(f"Invalid type {self.type_!r} for var declaration.")
+        for name in self.names:
+            if not _is_symbol_valid(name):
+                raise ValueError(f"Invalid name {name!r} for var declaration.")
+
+
+@dataclass(frozen=True)
+class SubroutineBody(AbstractSyntaxTree):
+    """Represents the body of a subroutine.
+
+    Attributes:
+        variables: The list of variable declarations.
+        statements: The statements inside the subroutine.
+    """
+
+    variables: list[VarDeclaration]
+    statements: Statements
+
+    def to_xml(self, doc: Document) -> Element:
+        element = doc.createElement("subroutineBody")
+        _add_children(element, "{", self.variables, self.statements, "}")
+        return element
+
+
+@dataclass(frozen=True)
+class SubroutineDeclaration(AbstractSyntaxTree):
+    """Represents a subroutine declaration.
+
+    Attributes:
+        type_: The type of the subroutine.
+        returns: The return type of the subroutine.
+        name: The name of the subroutine.
+        parameters: The parameters of the subroutine.
+        body: The body of the subroutine.
+    """
+
+    type_: str
+    returns: str
+    name: str
+    parameters: ParameterList
+    body: SubroutineBody
+
+    @property
+    def is_identifier(self) -> bool:
+        """True if return type is identifier."""
+        return self.returns not in {"int", "char", "boolean", "void"}
+
+    def to_xml(self, doc: Document) -> Element:
+        element = doc.createElement("subroutineDec")
+        _add_children(
+            element,
+            ("keyword", self.type_),
+            ("identifier" if self.is_identifier else "keyword", self.returns),
+            ("identifier", self.name),
+            "(",
+            self.parameters,
+            ")",
+            self.body,
+        )
+        return element
+
+    def __post_init__(self):
+        if not _is_symbol_valid(self.type_):
+            raise ValueError(f"Invalid type {self.type_!r} for subroutine declaration.")
+        if not _is_symbol_valid(self.returns):
```

Comment:
Fix self.returns check to check that it could be a valid symbol or allowed built in type.